File: Lua/BubbleSort.lua
--BubbleSort
function BubbleSort(arr) 
  for i=1,#arr do
    for j=1,#arr-i-1 do
      if(arr[j+1] < arr[j]) then
        Swap(arr,j+1,j);
      end
    end
  end
end

function Swap(arr,i,j)
  local temp = arr[i];
  arr[i] = arr[j];
  arr[j] = temp;
end

local test = {124,5,6,7326,1,3,26,236,74322}
BubbleSort(test);
for k,v in ipairs(test) do
  print(v);
end

File: C#/SortingAlgorithm/BubbleSort.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SortingAlgorithm
{
    public class BubbleSort
    {
        /// <summary>
        /// 普通冒泡排序
        /// </summary>
        public static void Sort(int[] arr)
        {
            for (int i = 0; i < arr.Length; i++)
            {
                for (int j = 0; j < arr.Length - i - 1; j++)
                {
                    if (arr[j] > arr[j + 1])
                    {
                        Util.Swap(arr, j, j + 1);
                    }
                }
            }
        }

        /// <summary>
        /// 鸡尾酒排序,定向冒泡排序,是冒泡排序的一种变形
        /// </summary>
        /// <param name="arr"></param>
        public static void CocktailSort(int[] arr)
        {
            bool swaped = true;
            int leftIndex = 0;
            int rightIndex = arr.Length - 1;

            while (swaped && rightIndex > leftIndex)
            {
                swaped = false;

                for (int j = leftIndex; j < rightIndex ; j++)
                {
                    if (arr[j] > arr[j + 1])
                    {
                        Util.Swap(arr, j, j + 1);
                        swaped = true;
                    }
                }
                rightIndex--;

                for (int j = rightIndex; j > leftIndex; j--)
                {
                    if (arr[j] < arr[j - 1])
                    {
                        Util.Swap(arr, j, j - 1);
                        swaped = true;
                    }
                }
                leftIndex++;
            }
        }
    }
}
